pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+	}
+}
+
+func (rg *csharpResourceGenerator) generateInputType(typ *propertyType, baseType string, nested, plainType bool) {
+	contract.Assert(typ.kind == kindObject)
+
+	rg.w.Writefmtln("")
+
+	// Open the class.
+	if baseType != "" {
+		baseType = " : " + baseType
+	}
+	rg.w.Writefmtln("    public class %s%s", typ.name, baseType)
+	rg.w.Writefmtln("    {")
+
+	// Declare each input property.
+	for _, p := range typ.properties {
+		rg.generateInputProperty(p, nested, plainType)
+		rg.w.Writefmtln("")
+	}
+
+	// Generate a constructor that will set default values.
+	rg.w.Writefmtln("        public %s()", typ.name)
+	rg.w.Writefmtln("        {")
+	for _, prop := range typ.properties {
+		if defaultValue := csharpDefaultValue(prop); defaultValue != "null" {
+			propertyName := csharpPropertyName(prop)
+			rg.w.Writefmtln("            %s = %s;", propertyName, defaultValue)
+		}
+	}
+	rg.w.Writefmtln("        }")
+
+	// Close the class.
+	rg.w.Writefmtln("    }")
+}
+
+func (rg *csharpResourceGenerator) generateOutputType(typ *propertyType, nested bool) {
+	contract.Assert(typ.kind == kindObject)
+
+	qualifier := ""
+	if !nested {
+		qualifier = "Outputs"
+	}
+
+	rg.w.Writefmtln("")
+
+	// Open the class and attribute it appropriately.
+	rg.w.Writefmtln("    [OutputType]")
+	rg.w.Writefmtln("    public class %s", typ.name)
+	rg.w.Writefmtln("    {")
+
+	// Generate each output field.
+	for _, prop := range typ.properties {
+		fieldName := csharpPropertyName(prop)
+		fieldType := qualifiedCSharpPropertyType(prop, qualifier, false /*wrapInput*/)
+		emitCSharpPropertyDocComment(rg.w, prop, "        ")
+		rg.w.Writefmtln("        public readonly %s %s;", fieldType, fieldName)
+	}
+	if len(typ.properties) > 0 {
+		rg.w.Writefmtln("")
+	}
+
+	// Generate an appropriately-attributed constructor that will set this types' fields.
+	rg.w.Writefmtln("        [OutputConstructor]")
+	rg.w.Writefmt("        private %s(", typ.name)
+
+	// Generate the constructor parameters.
+	for i, prop := range typ.properties {
+		paramName := csharpIdentifier(prop.name)
+		paramType := qualifiedCSharpPropertyType(prop, qualifier, false /*wrapInput*/)
+
+		terminator := ""
+		if i != len(typ.properties)-1 {
+			terminator = ","
+		}
+
+		paramDef := fmt.Sprintf("%s %s%s", paramType, paramName, terminator)
+		if len(typ.properties) > 1 {
+			paramDef = "\n            " + paramDef
+		}
+		rg.w.WriteString(paramDef)
+	}
+	rg.w.Writefmtln(")")
+
+	// Generate the constructor body.
+	rg.w.Writefmtln("        {")
+	for _, prop := range typ.properties {
+		paramName := csharpIdentifier(prop.name)
+		fieldName := csharpPropertyName(prop)
+		rg.w.Writefmtln("            %s = %s;", fieldName, paramName)
+	}
+	rg.w.Writefmtln("        }")
+
+	// Close the class.
+	rg.w.Writefmtln("    }")
+}
+
+func (rg *csharpResourceGenerator) generateInputTypes(nts []*csharpNestedType) {
+	// Open the Inputs namespace.
```

Comment:
note: i'm adding an item to a list wrt organization of our types in the .net namespace.  it makes a much bigger difference there since everything is nominal instead of structural.

Reply:
Tracking in https://github.com/pulumi/pulumi/issues/3406

Reply:
I think namespaces are good but we shouldn't add `Inputs` here. I basically import `Bla` and `Bla.Inputs` all the time. I would merge them.

Reply:
I needed this nested namespace because of input type names that overlapped with resource type names.

Reply:
Oh, this is bad... Do you remember an example of this?

Reply:
...now that I think about it, it may just have been output type names that overlapped, as those do not have a suffix. Would we be okay with either suffixing output type names or having an asymmetry between where input and output types reside (i.e. removing the `Inputs` namespace but keeping the `Outputs` namespace)?

Reply:
That makes sense. I didn't need to type the name of an output type a single time yet, so I may be less concerned about those.
The irrational sense of order doesn't like the asymmetry of suffix/subnamespace though... But I think it's a lesser evil. What do you think?
cc @CyrusNajmabadi

Reply:
i would be fine with this.

Reply:
note: i would just get something working now.  we can come back and decide on an orgnaizational and naming structure for .net in the future.  For example, i think we'll likely have to rethink the huge names we generate as tehy're much more impactful on the user experience in c# than they are in ts/js.

Reply:
Alright, I'll give that a go.

Reply:
Here's what happens in `pulumi-aws`:

```
Ec2/VpcPeeringConnectionAccepter.cs(128,25): error CS0101: The namespace 'Pulumi.Aws.Ec2' already contains a definition for 'VpcPeeringConnectionAccepterArgs' [/Users/pgavlin/dev/go/src/github.com/pulumi/pulumi-aws/sdk/dotnet/Pulumi.Aws.csproj]
Ec2/VpcPeeringConnectionAccepter.cs(168,16): error CS0111: Type 'VpcPeeringConnectionAccepterArgs' already defines a member called '.ctor' with the same parameter types [/Users/pgavlin/dev/go/src/github.com/pulumi/pulumi-aws/sdk/dotnet/Pulumi.Aws.csproj]
```

There is a resource--`VpcPeeringConnectionAccepter`--that generates a top-level `VpcPeeringConnectionAccepterArgs` type into the `Pulumi.Aws.Ec2` namespace. The `VpcPeeringConnection` type also has an `Accepter` field that generates a nested `VpcPeeringConnectionAccpterArgs` type. If these two types are generated into the same namespace, we get this conflict. So I think we need to keep the namespaces.

Reply:
A couple of minor things.  These "inputs" are always into resources, and we require them to derive from ResourceArgs.  As such, they probably should end with 'Args' all the time (if htey're not already doing that).  So `BucketArgs`, `VpcSomeSomePropArgs`, etc.  That would likely ensure that inputs don't collide without outputs.